server/stream.go

```diff
@@ -6067,18 +6080,35 @@ func (mset *stream) ackMsg(o *consumer, seq uint64) {
 	case InterestPolicy:
 		shouldRemove = mset.noInterest(seq, o)
 	}
-	mset.mu.Unlock()
 
 	// If nothing else to do.
 	if !shouldRemove {
-		return
+		mset.mu.Unlock()
+		return false
+	}
+
+	if !mset.isClustered() {
+		mset.mu.Unlock()
+		// If we are here we should attempt to remove.
+		if _, err := store.RemoveMsg(seq); err == ErrStoreEOF {
+			// This should not happen, but being pedantic.
+			mset.registerPreAckLock(o, seq)
+		}
+		return true
 	}
 
-	// If we are here we should attempt to remove.
-	if _, err := store.RemoveMsg(seq); err == ErrStoreEOF {
-		// This should not happen, but being pedantic.
-		mset.registerPreAckLock(o, seq)
+	// Only propose message deletion to the stream if we're consumer leader, otherwise all followers would also propose.
+	// We must be the consumer leader, since we know for sure we've stored the message and don't register as pre-ack.
+	if o != nil && !o.IsLeader() {
+		mset.mu.Unlock()
+		// Must still mark as removal if follower. If we become leader later, we must be able to retry the proposal.
+		return true
 	}
+
```

Comment:
Could o be nil here?

Reply:
Yes, for example in `mset.internalLoop`
```go
case <-amch:
	seqs := ackq.pop()
	for _, seq := range seqs {
		mset.ackMsg(nil, seq)
	}
	ackq.recycle(&seqs)
```

Reply:
I meant in terms of the condition asssuming we are leader.

Reply:
This can only happen in `AckNone` and interest/WQ. Nothing pushes to `mset.ackq` otherwise and nothing else calls `ackMsg(nil, ...)`.